Combine all migrations into one

januszm opened the pull request:
Currently, all migrations added to this project over time are also copied to the target application on:
`rake acts_as_taggable_on_engine:install:migrations`
This is unnecessary; such migrations modify each other's effects (adding or removing indexes) and clutter the target repository. This PR creates a single migration that is a composite of all previous ones.

januszm committed:
Combine all migrations into one

januszm commented:
Related to: https://github.com/mbleigh/acts-as-taggable-on/pull/1092
let me know if I should remove some indices from the combined migration

januszm commented:
Looking at the taggings table indices, there are several redundancies due to how composite indices work:

###  Redundant Indices:
1. ` add_index :tag_id` - REDUNDANT
  Already covered by the leftmost column of taggings_idx (line 29-31)
1. `add_index :taggable_id` - PARTIALLY REDUNDANT
  Covered by taggings_taggable_context_idx (line 33-34) and taggings_idy (line 42-43)
1. `add_index :taggable_type` - POTENTIALLY USEFUL
  Not covered by any composite index as a leftmost column
1. `add_index :tagger_id` - PARTIALLY REDUNDANT
  Covered by tagger_id, tagger_type composite index (line 41)
1. `add_index :context` - POTENTIALLY USEFUL
  Not leftmost in any composite index, but commonly queried alone

The t.references calls on lines 20-22 automatically create single-column indices:
 - index_taggings_on_tag_id
 - index_taggings_on_taggable  - composite on taggable_id, taggable_type
 - index_taggings_on_tagger  - composite on tagger_id, tagger_type


@seuros   if you like I can update it with a final, optimized migration that I used in my app after all, it has 6 indices instead of 11 while maintaining all necessary query patterns.

seuros commented:
You need to use defensive coding =>  If table exists, skip.  (existing users have their tables).

Migration version should be 7.1 not 6.0

No reversible migration. unless you want someone very angry at you when they upgrade the gem.

januszm commented:
You mean that some users are using a version of this gem that doesn't have one or few of the recent migrations? Yes, that makes sense, but on the other hand, for all new installations and new projects, it would be worth simplifying the process and simply copying one final migration. I think the best solution is to add this migration to the next major version release, informing the README that if you're on version <13, first update to 13. Generate the migrations, and only then update to 14. Anyone who performs a new installation on version 14 will already have the new migration file. But also not a big deal to keep it the way it is, this PR was meant as quality of life improvement for new installations

seuros commented:
No , i mean there is 25k project dependent on this gem. 

- I release any version (major or patch)
- they upgrade
- their IDE show new migration available.
- They run it
- Crash because they have the table since 2012.

Rails migration is basic... it see a new name, it add it to the application migration.

This is why when i rewrote this gem, i made it so the migrations are the responsibility of the user.

januszm commented:
Migrations are only copied when this task is invoked: `rake acts_as_taggable_on_engine:install:migrations`
but gotcha, updated the migration with  condition checks for skipping and also changed the down function to Irreversible unless someone wants to uncomment drop statements for new installations.

seuros commented:
> Migrations are only copied when this task is invoked: rake acts_as_taggable_on_engine:install:migrations

That correct, but if you use an IDE or AI assistance that try to be `helpful`, it will follow pattern to install migration once it see AATO gem update.

The changes look correct.

januszm commented on db/migrate/1_setup_acts_as_taggable_on.rb:
```diff
@@ -0,0 +1,66 @@
+    unless table_exists?(ActsAsTaggableOn.tags_table)
+      create_table ActsAsTaggableOn.tags_table do |t|
+        t.string :name
+        t.string :color
```
Looks like the color that I'm using in my app sneaked in, removed in most recent commit

januszm commented:
Removed color (althought it's worth adding in a separate migration, modern UI libraries for displaying tags use colors). Let's also apply all Copilot's suggestions, he's probably right

seuros commented:
> Removed color (althought it's worth adding in a separate migration, modern UI libraries for displaying tags use colors)

That not the concept of tags... Those are badges.  Tags are the values.
